### ISSUE TITLE
Timex.Ecto.Datetime magnificently slow decode times.

### ISSUE THREAD

**Issue by benwilson512** [bitwalker/timex_ecto#26]
Hey!

TL;DR: https://github.com/bitwalker/timex_ecto/blob/master/lib/types/datetime.ex#L33 is super slow. Timex.Ecto should probably consider just building a Timex.Ecto.DateTime struct manually.

I was recently trying to debug an issue in production where the decoding time of a mere ~1500 records was taking 400ms. I eliminated the time columns from the select and it dropped to 20ms. I switched to Calecto and with the time columns it's decoding in about 40ms, which is a 10x improvement.

Switching from Timex.Ecto to Calecto overall improved my test suite run time by 40%.

The primary reason as best I can tell is that Timex.Ecto calls Timex.datetime instead of directly creating a struct

**Comment by benwilson512** [comment]
This code here https://github.com/benwilson512/timex-bench shows about an 11x difference.

**Comment by benwilson512** [comment]
```
Calecto.DateTimeUTC.load tuple    10000000   0.13 µs/op
Calecto.DateTimeUTC.cast map        500000   6.50 µs/op
Timex.Ecto.Datetime.load tuple       50000   65.78 µs/op
Timex.Ecto.Datetime.cast map         50000   73.90 µs/op
```

**Comment by bitwalker** [comment]
I've got fixes locally now that bring Calendar and Timex up to roughly equivalent perf. Timex will still be a shade slower, but I'm ok with that because I'm performing extra checks in a lot of places that Calendar is not. In any case, a new release will be dropped sometime this week with the changes I've made. From what I gather you've already moved on to Calendar, but I still appreciate the report!

**Comment by benwilson512** [comment]
Thanks! Greatly appreciated :)

**Comment by bitwalker** [comment]
This should be addressed in the latest release, I'm going to close this for now, but please reopen it if you are still seeing performance issues. Thanks again for the report!
